### PR TITLE
Decouple system timer and drivers: I2C

### PR DESCRIPTION
**Pull request by pizi-nordic**
During our work in timer area, we found that system clock frequency is used incorrectly in several drivers.
For example, it is used to obtain bus frequency needed to calculate speed of I2C interfaces.
If we select different system timer device, all these calculations become invalid.

This PR fixes the problem by obtaining all needed information directly from the DTS.
Brings us closer to: #15363

Please note, that for I2C devices the clock-frequency property specifies SCK frequency, instead of frequency of the clock driving  peripheral. To solve that problem, a new property was added.
(I used https://www.kernel.org/doc/Documentation/devicetree/bindings/i2c/i2c-ocores.txt as a reference)

### CODE REVIEW

**Review by ioannisg — COMMENTED** [review]
Although, I don't disagree with the change (and in general with this kind of patches), I wonder if the "clock-frequency" is actually going to be "copied" in all DTS driver nodes, for each platform, e.g. SPIs, UARTs, etc.

If, so, I wonder if there a way to "get" the value of this clock-frequency, somehow, and not hard-code it in every peripheral node. I understand that different peripherals might have different clock sources, so until we have a proper clock management, this, all , might be hard, so I am just wondering.

**Comment by pizi-nordic** [comment]
We should have a "clock" node which references the clock source for the peripheral. But this needs both entries in DTS as well as better DTS infrastructure (ideally we should have something like DT_I2C_0_CLOCK_0_FREQUENCY extracted from clock node.).

File: dts/bindings/i2c/ti,cc32xx-i2c.yaml [hunk 1]
```diff
@@ -18,6 +18,12 @@ properties:
       generation: define
       category: required
 
+    "cc32xx,clock-frequency" :
```
Comment (anangl) [inline dts/bindings/i2c/ti,cc32xx-i2c.yaml]: What about naming it `ip-clock-frequency`? And consistently using `IP_` also in the fixup label? `DT_I2C_0_CLOCK_FREQUENCY` looks like something generated for the `clock-frequency` property.

Comment (pizi-nordic) [reply dts/bindings/i2c/ti,cc32xx-i2c.yaml]: No problem for me. Any other opinions/suggestions?

Comment (carlescufi) [reply dts/bindings/i2c/ti,cc32xx-i2c.yaml]: @anangl why `ip-`?

**Comment by carlescufi** [comment]
> We should have a "clock" node which references the clock source for the peripheral. But this needs both entries in DTS as well as better DTS infrastructure (ideally we should have something like DT_I2C_0_CLOCK_0_FREQUENCY extracted from clock node.).

I agree that this is the real solution, as shown in examples like this one:
https://www.kernel.org/doc/Documentation/devicetree/bindings/i2c/i2c-stm32.txt

File: dts/bindings/i2c/ti,cc32xx-i2c.yaml (hunk 1 above)
Comment (carlescufi) [reply dts/bindings/i2c/ti,cc32xx-i2c.yaml]: the problem with `IP` is that it really reminds of IP block, and this is not the case. So I would instead either:

1. Go with `<name>,clock-frequency`
2. Use something like `<name>,hw-clock-frequency`

@erwango @galak thoughts?

File: dts/bindings/i2c/ti,cc32xx-i2c.yaml [hunk 2]
```diff
@@ -18,6 +18,12 @@ properties:
...
+      type: int
+      category: required
+      description: Clock frequency the SPI peripheral is being driven at
```
Comment (galak) [inline dts/bindings/i2c/ti,cc32xx-i2c.yaml]: the description is wrong.

File: dts/bindings/i2c/ti,cc32xx-i2c.yaml (hunk 1 above)
Comment (galak) [reply dts/bindings/i2c/ti,cc32xx-i2c.yaml]: We did something on sifive, looks like we used `input-frequency`.  But I'm not keen on propagating that everywhere

**Review by galak — CHANGES_REQUESTED** [review]
Why are we doing this via DTS?  Why isn't this some inline function?

**Comment by galak** [comment]
@pizi-nordic was the thread of discussion about how to handle 'clock' vs 'clock-frequency' for i2c here or somewhere else?

**Comment by pizi-nordic** [comment]
@galak: Here: https://github.com/zephyrproject-rtos/zephyr/pull/15415#discussion_r274844892.

**Pull request closed without merging**